### DISCUSSION TITLE
Question About Ant Behavior

### DISCUSSION THREAD

**Discussion by SumeetBatra** [google/brax#277] [General]
Hi folks, I have a question on how the ant environment behaves initially. I'm training RL policies with 2 layer mlp's using PPO and noticed that the initial rewards become quite negative before the policy begins to learn. I understand that this could be due to a myriad of differences in my PPO implementation, hyperparameters, model architecture etc. However, when I visualize just a randomly initialized policy, I see that sometimes ant flips over and accumulates large negative rewards up until timeout termination. Here's a screenshot that visualizes what's happening.

![ant_flipped_over](https://user-images.githubusercontent.com/14263936/207979124-60727a7c-d2df-4912-acc6-55104ac58425.png)

Is this correct behavior on the environment side? I would have thought there would be some termination condition if the ant flips over like that so that this doesn't continue until timeout termination. Or maybe I'm missing something.

**Comment by btaba** [comment]
Hi @SumeetBatra the environment should terminate when `terminate_when_unhealthy` is True (default), see https://github.com/google/brax/blob/main/brax/envs/ant.py#L241. However, if you're using something like the AutoResetWrapper, the environment will get reset automatically. Perhaps your random initialization is flipping the Ant, and the env keeps getting reset to the unhealthy state?

**Reply by SumeetBatra** [reply]
Hi @btaba, my evaluation loop looks like this: 
```
    while not done:
        with torch.no_grad():
            obs = obs.unsqueeze(dim=0).to(device)
            if cfg.normalize_obs:
                obs = (obs - obs_mean) / torch.sqrt(obs_var + 1e-8)
            act, _, _ = agent.get_action(obs)
            act = act.squeeze()
            obs, rew, done, info = env.step(act.cpu())
            rollout.append(env.unwrapped._state)
            total_reward += rew
```
So even with the AutoResetWrapper I don't believe this rollout includes autoresets because I exit on the first done=True flag

**Reply by btaba** [reply]
Ah I see, can you confirm what the qp.pos[0, 2] is (the height of the torso)? It's obs[2] I believe depending on how you set `exclude_current_positions_from_observation` (see here https://github.com/google/brax/blob/main/brax/envs/ant.py#L261-L265)
Can you also share a full example of how you create the environment?

Keep in mind that there is a negative control cost reward which you might be hitting: https://github.com/google/brax/blob/main/brax/envs/ant.py#L245
You can try inspecting the info to see how the reward is constructed leading to large negative rewards

**Reply by SumeetBatra** [reply]
I actually checked this right after my previous post and indeed the control cost coefficient was too high! This makes me wonder, are the default reward coefficients not used in the collab PPO example? I don't see the reward dip below 0 during the initial stages of training, though I suppose this could also be differences in logging. 

As for the issue where ant flips over, I can't seem to reproduce this behavior today for whatever reason and I have made no changes since the original post :disappointed: So perhaps it's not the most pressing issue, although I appreciate you looking into this for me!

**Reply by btaba** [reply]
Hi @SumeetBatra ! If you happen to get a repro for the Ant flipping and not resetting, please send it over because that will definitely be an issue to fix.

The default reward coefficients are all set in ant.py, except for `use_contact_forces` in https://github.com/google/brax/blob/main/brax/envs/__init__.py#L46.  Maybe your PPO impl starts off with actions with higher magnitude?

**Reply by SumeetBatra** [reply]
Okay, after many attempts, I was able to reproduce it. Here's a short video with the torso's position and rotation visualized 

https://user-images.githubusercontent.com/14263936/208753860-e2ea94a3-3698-4ecf-9415-cc993c85289f.mp4

It looks like the z-pos never goes outside the healthy z-range, so perhaps that's the bug?

**Reply by SumeetBatra** [reply]
and here's how I create my environment: 
```
def make_vec_env_brax(cfg):
    entry_point = functools.partial(brax_custom.create_gym_env, env_name=cfg.env_name)
    brax_env_name = _to_custom_env[cfg.env_name]['custom_env_name']
    if brax_env_name not in gym.envs.registry.env_specs:
        gym.register(brax_env_name, entry_point=entry_point)

    act_bounds = _to_custom_env[cfg.env_name]['action_clip']
    obs_bounds = _to_custom_env[cfg.env_name]['obs_clip']
    rew_bounds = _to_custom_env[cfg.env_name]['reward_clip']
    vec_env = gym.make(_to_custom_env[cfg.env_name]['custom_env_name'], batch_size=cfg.env_batch_size, seed=cfg.seed,
                       clip_actions=act_bounds, clip_rewards=rew_bounds, clip_obs=obs_bounds)
    vec_env = to_torch.JaxToTorchWrapper(vec_env, device='cuda')

    return vec_env
```

Which then calls 
```
def create(env_name: str,
           episode_length: int = 1000,
           action_repeat: int = 1,
           clip_actions: Optional[tuple] = None,
           clip_rewards: Optional[tuple] = None,
           clip_obs:     Optional[tuple] = None,
           auto_reset: bool = True,
           batch_size: Optional[int] = None,
           eval_metrics: bool = False,
           **kwargs) -> Env:
    """Creates an Env with a specified brax_custom system."""
    env = _envs[env_name](legacy_spring=True, **kwargs)
    env = FeetContactWrapper(env, env_name)
    if clip_obs:
        env = ObservationClipWrapper(env, obs_min=clip_obs[0], obs_max=clip_obs[1])
    if clip_rewards:
        env = RewardClipWrapper(env, rew_min=clip_rewards[0], rew_max=clip_rewards[1])
    if clip_actions:
        env = ActionClipWrapper(env, a_min=clip_actions[0], a_max=clip_actions[1])
    if episode_length is not None:
        env = wrappers.EpisodeWrapper(env, episode_length, action_repeat)
    if batch_size:
        env = wrappers.VectorWrapper(env, batch_size)
    if auto_reset:
        env = wrappers.AutoResetWrapper(env)
    if eval_metrics:
        env = wrappers.EvalWrapper(env)

    return env  # type: ignore
```
Actions are clipped (-1, 1) and obs/rew are clipped (-10, 10)

**Reply by btaba** [reply]
Hi @SumeetBatra. Ah got it, thanks for debugging! Indeed it looks like z doesn't go below 0.2. This is a difference with the Mujoco Ant env and Brax v1, where Mujoco only has contact with the [feet and the floor](https://github.com/google/brax/blob/main/brax/v2/envs/assets/ant.xml#L85-L90), not the torso and the floor. This is fixed in Brax v2 since we load directly from the Mujoco XML; I encourage you to check it out [here](https://github.com/google/brax/tree/main/brax/v2) and [here](https://colab.research.google.com/github/google/brax/blob/main/notebooks/Brax_v2_Training_Preview.ipynb).

**Reply by SumeetBatra** [reply]
Okay thanks for the update! Is there a rough timeline on when we can expect an official release of Brax v2? Also are other envs supported in the preview?

**Reply by btaba** [reply]
Hi @SumeetBatra, for a very rough timeline: in the next few months or so! Feel free to pin v0.0.16 or work off of v2, we'll eventually move all the v2 stuff into the root src folder.
We'll be adding all of the same envs as in v1 bit by bit, but we only have Ant for now
